deploy chi@edge v2

msherman64 opened the pull request:
This PR allows for deployment of a Kubernetes control plane via K3s onto all control nodes.
Additional worker nodes can be added as per https://rancher.com/docs/k3s/latest/en/quick-start/
Zun will be configured to use said K3s cluster to deploy containers onto these worker nodes.

Note: You'll need to deploy in the following sequence:
1. Run a full deploy with `enable_k3s` set to false.
2. Set `enable_k3s` to true.
3. Run `cc-ansible --playbook ./playbooks/k3s.yaml`
4. Run `cc-ansible deploy` again. The second pass is needed because zun depends on having a token for the k3s cluster.

Commit:
Update default network policies

Commit:
download calico release

Commit:
Support device plugins and neutron bridge

Copy the file out of git so it doesn't register a change when it
reverts the line replacement.

Commit:
fixup neutron config

Commit:
register device profiles

Include device manager files explicitly
It was too much effort needed to adjust these from the source repo, just vendor them.

Commit:
ensure k3s starts w/ secrets encryption

Commit:
handle addr already assigned

Commit:
update calico policies

Commit:
always apply both policies

Commit:
use calicoctl

Commit:
add vcsm-cma to pi_camera profile

Commit:
add support for device workers

Commit:
add balena api token secret

Commit:
enable blazar k8s plugin

Commit:
add toggle for network plugin

Commit:
fix with learnings from calico

Commit:
loosen wireguard interface detect

tunelo ports have wg interfaces that have part of the port uuid in
the name, so we can't have a fixed pattern here.

Commit:
update fleet name

Commit:
fixup neutron connect script

Commit:
bind on to non-vip

Commit:
add config for k8s worker

msherman64 committed:
update label syntax to enable smarterdevice

the smarterdevice daemonset looks for the label `smarter-device-manager=enabled`,
 rather than  `smarter-device-manager/enabled=true`. Update the config in doni.conf to comply.

msherman64 committed:
add support for jetson xavier nx

Add support for jetson xavier nx. This adds the xavier to its own fleet,
as well as enabling smarter-device-manager.

Commit:
update subnet sizes

Commit:
Sync changes from prod

Commit:
Add docker registry proxy deployment

the registry proxy will be deployed to the control node exclusively.

msherman64 committed:
fixup paths and permissions

msherman64 committed:
move to use of kubernetes collection for setup

msherman64 committed:
fix selector for controlplane

msherman64 committed:
bump version of k3s plugins

msherman64 committed:
ensure k3s service uses consistent IP

msherman64 committed:
deploy nvidia device plugins with k8s module

msherman64 committed:
refactor k3s service and client install

msherman64 committed:
ensure zun can start with xena release

msherman64 commented:
This has some remaining issues before it can be merged.
1. conflicts in ml2_conf.ini, blazar.conf overrides
2. k3s playbook depends on post-deploy having run, but zun deploy won't finish without a kubernetes token, generated by k3s playbook

Commit:
add k3s playbook

msherman64 commented on kolla/node_custom_config/doni.conf:
```diff
@@ -1,5 +1,8 @@
 [DEFAULT]
-enabled_worker_types = ironic{% if enable_blazar | bool %},blazar.physical_host{% endif %}
+{# Note: newlines after these options are important so jinja doesn't squash them #}
+enabled_hardware_types = {% if enable_ironic | bool %}baremetal,{% endif %}{% if enable_k3s | bool %}device.balena,{% endif %}
```
Having the logic in this template is getting very messy, and seems incorrect on this line. We need a better way to print a "list" of enabled drivers, without a chain of if statements.

diurnalist replied:
What part seems incorrect? It looks OK to me. Also for some reason this comment doesn't show up in the diff view 🤔 

Yeah it's not super readable. One long term architectural option would be to have workers disable themselves under some condition, e.g., if Blazar is not in the service catalogue, the Blazar worker never "turns on" when Doni starts its worker manager.

I'm not sure of a better way to template all that stuff though. Ansible tries to not give you too many control flow constructs to avoid templates being turing complete :) - that applies to stuff like tasks as well. It might require an additional filter, which seems like overkill.

msherman64 replied:
Oops, I selected the wrong line for the comment, misclick. Making a new comment with the correct line.

msherman64 commented on kolla/node_custom_config/neutron/ml2_conf.ini:
```diff
@@ -2,11 +2,10 @@
+# Allow creating flat networks against any physical provider network
+# (in practice this isn't always valid, but Kolla Ansible by default is opinionated
+# about what the provider networks should be named.)
 flat_networks = *
```
This is just a "loosening" on the restrictions for what physnets can hold a network without a segmentation ID. Only admins can create such networks, so this is probably fine to leaver as `*`

msherman64 commented on kolla/node_custom_config/zun.conf:
```diff
@@ -7,15 +7,34 @@ minimum_disk = 0
+{% if enable_zun_compute_k8s | bool %}
+# We support containers, but not capsules when k8s driver is in use.
+container_driver = k8s
+capsule_driver = fake
```
This needs to be set as "fake" until the k8s driver advertises support for capsules, even if a no-op.

msherman64 commented on kolla/node_custom_config/zun.conf:
```diff
@@ -7,15 +7,34 @@ minimum_disk = 0
 [scheduler]
 available_filters = zun.scheduler.filters.all_filters
-enabled_filters = {% if enable_blazar | bool %}BlazarFilter,{% endif %}RamFilter,CPUFilter,ComputeFilter,RuntimeFilter
+enabled_filters = {% if enable_blazar | bool %}BlazarFilter,{% endif %}ComputeFilter,RuntimeFilter
```
RamFilter and CPUFilter are not supported in Xena, got merged with placement.

msherman64 commented on requirements.txt:
```diff
@@ -3,6 +3,7 @@ ansible>=2.10,<=4
 # See https://jinja.palletsprojects.com/en/3.1.x/changes/#version-3-1-0
 jinja2<3.1.0
 docker
+kubernetes
```
This is a dependency of the kubernetes ansible collection.

msherman64 reviewed (commented):
Some changes are still needed before merging, but I could use some feedback on the best way to do so. In particular the templating in doni.conf.

diurnalist commented on kolla/defaults.yml:
```diff
@@ -310,9 +311,14 @@ enable_grafana: no
+blazar_enable_device_plugin_k8s: "{{ enable_k3s | bool }}"
+# If not running K3s, can run standard Zun deploy w/ etcd and kuryr
+enable_etcd: "{{ enable_zun | bool and not enable_k3s | bool }}"
+enable_kuryr: "{{ enable_zun | bool and not enable_k3s | bool }}"
```
We may not even want to imply that traditional Zun is deployable since we are not going to be testing it at all going forward.

diurnalist commented on roles/k3s/tasks/config-neutron.yml:
```diff
@@ -0,0 +1,51 @@
+      enable_dhcp: no
+  become: yes
+
+# FIXME(jason): this doesn't actually add the subnet to the router!
```
Just highlighting this FIXME, maybe in Xena it's fixable b/c Ansible is at a later version.

msherman64 replied:
I actually have a fix for this in some later commits! 
using the openstack.cloud collection

* [subnets_info](https://docs.ansible.com/ansible/latest/collections/openstack/cloud/subnets_info_module.html#ansible-collections-openstack-cloud-subnets-info-module)
* [subnet](https://docs.ansible.com/ansible/latest/collections/openstack/cloud/subnet_module.html#ansible-collections-openstack-cloud-subnet-module)
* [routers_info](https://docs.ansible.com/ansible/latest/collections/openstack/cloud/routers_info_module.html#ansible-collections-openstack-cloud-routers-info-module)
* [routers](https://docs.ansible.com/ansible/latest/collections/openstack/cloud/router_module.html#ansible-collections-openstack-cloud-router-module)

diurnalist commented on roles/k3s/templates/neutron-calico-connect.j2:
```diff
@@ -0,0 +1,37 @@
+#!/usr/bin/env bash
```
the memories...

msherman64 replied:
technically calico seems to support floating IPs natively, but not in the operator deployed version. Could be interesting for future work.
https://projectcalico.docs.tigera.io/networking/add-floating-ip#enable-floating-ips

msherman64 committed:
Refactor templating logic for clarity

This refactors the templating logic for doni workers and hardware types.
Appending to a list on condition, then joining the list, separates the
logic from syntax.

However, it may be preferrable move the logic to an ansible set_facts
role instead, as logic in a jinja template is an antipattern.
